src/vec.rs

```diff
@@ -0,0 +1,296 @@
+        let heap = Self {
+            memory,
+            max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: 0,
+            _phantom: PhantomData,
+        };
+
+        heap.save();
+        heap
+    }
+
+    pub fn load(memory: M) -> Self {
+        Self::load_with_sizes(memory, T::max_size() as u64)
+    }
+
+    pub fn load_with_sizes(memory: M, max_value_size: u64) -> Self {
+        // Read the header from memory.
+        let header: StableVecHeader = read_struct(Address::from(0), &memory);
+        assert_eq!(&header.magic, MAGIC, "Bad magic.");
+        assert_eq!(header.version, LAYOUT_VERSION, "Unsupported version.");
+        let expected_value_size = header.max_value_size;
+        assert!(
+            max_value_size <= expected_value_size,
+            "max_value_size must be <= {}",
+            expected_value_size
+        );
+
+        Self {
+            memory,
+            max_value_size: header.max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: header.len,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn save(&self) {
+        let header = StableVecHeader {
+            magic: *MAGIC,
+            version: LAYOUT_VERSION,
+            max_value_size: self.max_value_size,
+            len: self.len,
+            _buffer: [0; 24],
+        };
+
+        write_struct(&header, Address::from(0), &self.memory);
+    }
+
+    fn index_to_addr(&self, index: usize) -> Address {
+        self.base + Bytes::from(self.max_value_size * (index as u64))
+    }
+}
+
+impl<M: Memory + Clone, T: BoundedStorable> StableVec<M, T> {
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        let page_size = 64 * 1024;
+        let bytes = self.memory.size() * page_size;
+        let free_bytes = bytes - StableVecHeader::size().get();
+        (free_bytes / self.max_value_size) as usize
+    }
+
+    pub fn reserve(&mut self, additional: usize) {
+        let page_size = 64 * 1024;
+        let round_up = |x| ((x + page_size - 1) & !(page_size - 1));
+
+        let add_bytes = additional as u64 * self.max_value_size;
+        let add_pages = round_up(add_bytes) / page_size;
+
+        self.memory.grow(add_pages);
+    }
+
+    pub fn insert(&mut self, index: usize, element: &T) {
+        #[cold]
+        #[inline(never)]
+        fn assert_failed(index: usize, len: usize) -> ! {
+            panic!("insertion index (is {index}) should be <= len (is {len})");
+        }
+
+        let len = self.len();
+        if index > len {
+            assert_failed(index, len);
+        }
+
+        // space for the new element
+        if len == self.capacity() {
+            self.reserve(1);
+        }
+
+        let p = self.index_to_addr(index);
+        let src = p;
+        let dst = p + Bytes::from(self.max_value_size as u64);
+        copy::<T, M>(&self.memory, src, dst, (len - index) as u64);
```

Comment:
I'm not sure I understand what we're doing here.

Reply:
Well, the default behavior of insert is that it will add an entry at position 'index'.
so if v = [0,1,2,3]
and we do insert(1, 10)
we get v = [0,10,1,2,3].

For this we need to move [1,2,3] 1 over to the right. Which is what we do here.

I realize that this code is not very self explanatory, so i'll add a comment.